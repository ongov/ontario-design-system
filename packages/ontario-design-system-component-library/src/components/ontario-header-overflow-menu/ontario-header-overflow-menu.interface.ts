export interface MenuItem {
	title: string;
	href: string;
	linkIsActive?: boolean;
	onClickHandler?: (event: Event) => void;
}
