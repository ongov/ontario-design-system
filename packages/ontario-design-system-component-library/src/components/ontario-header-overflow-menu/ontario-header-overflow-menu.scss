@use 'sass:math';
@use '@ongov/ontario-design-system-global-styles/dist/styles/scss/1-variables/global.variables' as globalVariables;
@use '@ongov/ontario-design-system-global-styles/dist/styles/scss/1-variables/spacing.variables' as spacing;
@use '@ongov/ontario-design-system-global-styles/dist/styles/scss/1-variables/colours.variables' as colours;
@use '@ongov/ontario-design-system-global-styles/dist/styles/scss/1-variables/breakpoints.variables' as breakpoints;
@use '@ongov/ontario-design-system-global-styles/dist/styles/scss/1-variables/fonts.variables' as fonts;
@use '@ongov/ontario-design-system-global-styles/dist/styles/scss/2-tools/functions/global.functions' as globalFunctions;
@use '@ongov/ontario-design-system-global-styles/dist/styles/scss/1-variables/font-weights.variables' as fontWeights;
@use '@ongov/ontario-design-system-global-styles/dist/styles/scss/2-tools/placeholder/focus.placeholders' as
	focusPlaceholders;
@use '@ongov/ontario-design-system-global-styles/dist/styles/scss/1-variables/z-index.variables' as zIndex;
@use '@ongov/ontario-design-system-global-styles/dist/styles/scss/1-variables/typography.variables';
@use '@ongov/ontario-design-system-global-styles/dist/styles/scss/5-layout/_grid.layout.scss';
@use '@ongov/ontario-design-system-global-styles/dist/styles/scss/6-components/_text-inputs.component.scss';
@use '@ongov/ontario-design-system-global-styles/dist/styles/scss/7-overrides/_visibility.overrides.scss';
@use '../../styles/header.scss';

$navigation-link-bg--hover: #e5e5e5;

.ontario-application-navigation {
	@extend %ontario-application-navigation;

	ul {
		list-style: none;
		padding: spacing.$spacing-2 spacing.$spacing-0;
		margin: spacing.$spacing-0;
		border-bottom: 1px solid colours.$ontario-greyscale-20;

		@media screen and (min-width: breakpoints.$medium-breakpoint) {
			border-bottom: none;
		}
	}

	li {
		border-top: 1px solid colours.$ontario-greyscale-20;
		padding: spacing.$spacing-0;
		margin-bottom: spacing.$spacing-0;

		&:last-of-type {
			border-bottom: 1px solid colours.$ontario-greyscale-20;
		}
	}

	a {
		color: colours.$ontario-colour-black;
		display: block;
		font-size: 1.125rem;
		font-weight: fontWeights.$ontario-font-weights-semi-bold;
		font-family: fonts.$ontario-font-open-sans;
		text-decoration: none;
		padding: calc(spacing.$spacing-3 - math.div(spacing.$spacing-1, 3.25)) spacing.$spacing-5;

		&:hover {
			text-decoration: underline;
			background-color: $navigation-link-bg--hover;
		}

		&:active {
			text-decoration: underline;
			background-color: colours.$ontario-greyscale-20;
		}

		&:focus,
		&:active {
			@extend %ontario-focus;
			box-shadow: focusPlaceholders.$ontario-focus-box-shadow inset;
		}

		&.ontario-link--active {
			border-left: 6px solid colours.$ontario-colour-black;
			background-color: colours.$ontario-greyscale-5;
		}
	}
}

.ontario-application-navigation__container {
	@extend %ontario-application-navigation__container;
}

.ontario-application-navigation.ontario-navigation--open .ontario-application-navigation__container {
	background: colours.$ontario-colour-white;
	visibility: visible;
	transform: translateY(0);
	transition: transform 0.4s ease-in-out;

	@media screen and (min-width: breakpoints.$medium-breakpoint) {
		transition: none;
		display: block;
	}
}

.ontario-application-navigation.ontario-navigation--open {
	background: colours.$ontario-colour-white;
	z-index: zIndex.$ontario-z-index-above-high;
}
