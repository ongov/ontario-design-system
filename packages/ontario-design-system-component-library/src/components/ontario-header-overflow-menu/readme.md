import { OntarioHeaderOverflowMenu } from '@ongov/ontario-design-system-component-library-react';
import useBaseUrl from '@docusaurus/useBaseUrl';

# ontario-header-overflow-menu

This component is intended to be used as a sub-component of the ontario-header component.

## Configuration

Once the component package has been installed (see Ontario Design System Component Library for installation instructions), the ontario-header-overflow-menu component can be added directly into the project's code, and can be customized by updating the properties outlined [here](#properties). Please see the [examples](#examples) below for how to configure the component.

## Examples

```html
<ontario-header-menu-overflow
	menu-items='[
    {
      "title":"Item 1",
      "href":"/item-1"
    },
    {
      "title":"Item 2",
      "href":"/item-2"
    },
    {
      "title":"Item 3",
      "href":"/item-3"
    },
    {
      "title":"Item 4",
      "href":"/item-4"
    },
  ]'
	trap-menu-focus="false"
>
</ontario-header-menu-overflow>
```

<div>
    <OntarioHeaderMenuOverflow
    menu-items='[
      {
        "title":"Item 1",
        "href":"/item-1"
      },
      {
        "title":"Item 2",
        "href":"/item-2"
      },
      {
        "title":"Item 3",
        "href":"/item-3"
      },
      {
        "title":"Item 4",
        "href":"/item-4"
      },
    ]'
    trap-menu-focus='false'
  >
  </OntarioHeaderMenuOverflow>
</div>

<!-- Auto Generated Below -->

## Properties

| Property    | Attribute    | Description                             | Type                   | Default     |
| ----------- | ------------ | --------------------------------------- | ---------------------- | ----------- |
| `menuItems` | `menu-items` | The items that will go inside the menu. | `MenuItem[] \| string` | `undefined` |

## Dependencies

### Used by

- [ontario-header](../ontario-header)

### Graph

```mermaid
graph TD;
  ontario-header --> ontario-header-overflow-menu
  style ontario-header-overflow-menu fill:#f9f,stroke:#333,stroke-width:4px
```

---

_Built with [StencilJS](https://stenciljs.com/)_
