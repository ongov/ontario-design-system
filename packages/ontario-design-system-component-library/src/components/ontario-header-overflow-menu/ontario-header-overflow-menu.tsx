import { Component, Prop, State, Watch, h, Listen, Element } from '@stencil/core';
import { MenuItem } from './ontario-header-overflow-menu.interface';

@Component({
	tag: 'ontario-header-overflow-menu',
	styleUrl: 'ontario-header-overflow-menu.scss',
	shadow: true,
})
export class OntarioHeaderApplicationMenu {
	@Element() el: HTMLElement;

	/**
	 * The items that will go inside the menu.
	 *
	 *  @example
	 * 	<ontario-header-overflow-menu
	 *			menu-items='[{
	 *				"title": "Link 1",
	 *				"href": "/link-1"
	 *				"linkIsActive": "false"
	 *			},{
	 *				"title": "Link 2",
	 *				"href": "/link-2"
	 *				"linkIsActive": "false"
	 *			},{
	 *				"title": "Link 3",
	 *				"href": "/link-3"
	 *				"linkIsActive": "false"
	 *			},{
	 *				"title": "Link 4",
	 *				"href": "/link-4"
	 *				"linkIsActive": "false"
	 *			}]'>
	 *	</ontario-header-overflow-menu>
	 */
	@Prop() menuItems: MenuItem[] | string;

	/**
	 * The `menuItems` are reassigned to `menuItemState` for parsing by `parseMenuItems()`.
	 */
	@State() private menuItemState: MenuItem[];

	/**
	 * A state variable to determine whether the menu is open or not.
	 *
	 * Triggered by either the `menuButtonToggled` event or the `click` event.
	 */
	@State() private menuIsOpen: boolean = false;

	/**
	 * The current index of the menu item that is focused.
	 */
	private currentIndex: number | undefined = undefined;

	@Watch('menuItems')
	parseMenuItems() {
		if (!Array.isArray(this.menuItems) && typeof this.menuItems === 'string') {
			this.menuItemState = JSON.parse(this.menuItems);
		} else {
			this.menuItemState = this.menuItems;
		}

		// if a menu item is the active page, update the linkIsActive value for that menu item
		this.menuItemState.map((item) => {
			const activeLinkRegex = item.href.replace(/\s+/g, '-').toLowerCase();
			item.linkIsActive = window.location.pathname.includes(activeLinkRegex);
		});
	}

	/**
	 * This listens for the `menuButtonToggled` event sent from the header menu button when it is clicked.
	 * It is used to toggle menu visibility by adding or removing the ontario-navigation--open class on the nav element.
	 */
	@Listen('menuButtonToggled', { target: 'window' })
	toggleMenuVisibility(event: CustomEvent<boolean>) {
		this.menuIsOpen = event.detail;
		this.currentIndex = undefined;
	}

	/**
	 * Listens for clicks on anything outside of the menu and closes
	 * the menu by removing the ontario-navigation--open class on the nav element.
	 */
	@Listen('click', { capture: true, target: 'window' })
	handleClick(event: any) {
		// if the menu (ref) is clicked, return
		if (event.composedPath().includes(this.menu)) {
			this.menuIsOpen = true;
			return;
		}

		this.menuIsOpen = false;
		this.currentIndex = undefined;
	}

	/**
	 * Listens for keydown events and handles navigation within the menu using the ArrowUp and ArrowDown keys.
	 */
	@Listen('keydown', { target: 'window' })
	handleKeyDown(event: KeyboardEvent) {
		if (this.menuIsOpen) {
			const focusableElements: NodeListOf<HTMLElement> = this.menu.querySelectorAll('.ontario-menu-item');

			switch (event.key) {
				case 'ArrowDown':
					event.preventDefault();
					if (this.currentIndex === -1 || this.currentIndex === undefined) {
						this.currentIndex = 0;
					} else {
						this.currentIndex = (this.currentIndex + 1) % focusableElements.length;
					}
					focusableElements[this.currentIndex].focus();
					/**
					 * Tried other ways of triggering this function to condense code via a watch on currentIndex,
					 * and also tried sticking at the bottom of the switch's parent if statement.
					 *
					 * Both attempts led to inconsistent results (e.g. the aria message was out of sync with currentIndex)
					 * or just didn't appear at all. Setting this at the source seems to have the best outcome.
					 */
					this.updateAriaLive(this.currentIndex);
					break;
				case 'ArrowUp':
					event.preventDefault();
					if (this.currentIndex === -1 || this.currentIndex === undefined) {
						this.currentIndex = focusableElements.length - 1;
					} else {
						this.currentIndex = (this.currentIndex - 1 + focusableElements.length) % focusableElements.length;
					}
					focusableElements[this.currentIndex].focus();
					this.updateAriaLive(this.currentIndex);
					break;
				case 'Tab':
					focusableElements.forEach((element, index) => {
						element.addEventListener('focus', () => {
							this.currentIndex = index;
							this.updateAriaLive(this.currentIndex);
						});
					});
					break;
			}
		}
	}

	/**
	 * Updates the aria-live region with the menu item count and selected option.
	 *
	 * @param {number} selectedIndex The index of the selected menu item
	 */
	private updateAriaLive(selectedIndex: number) {
		const menuItemCount = this.menuItemState.length;
		const selectedMenuItemNumber = selectedIndex + 1;
		const ariaLiveMessage = `Option ${selectedMenuItemNumber} of ${menuItemCount}`;

		if (this.ariaLiveRegion) {
			this.ariaLiveRegion.textContent = ariaLiveMessage;
		}
	}

	/**
	 * This function generates the menu items in a <li>, accordingly, to the given parameters.
	 *
	 * href and title are necessary, but rest are not.
	 *
	 * @param href - the href of the menu item
	 * @param title - the title of the menu item
	 * @param linkIsActive - when set to true, this will add the classes necessary to style the link in a way that indicates to the user what the active page/link is
	 * @param liClass - if there is a class that is related to the <a> portion of the menu item, put it here
	 * @param onClick - for any custom onClick event a user might want to add to their menu links
	 */
	private generateMenuItem(
		href: string,
		title: string,
		linkIsActive: boolean | undefined,
		liClass?: string,
		onClick?: any,
	) {
		return (
			<li class={liClass}>
				<a class={`ontario-menu-item ${linkIsActive ? 'ontario-link--active' : ''}`} href={href} onClick={onClick}>
					{title}
				</a>
			</li>
		);
	}

	componentWillLoad() {
		this.parseMenuItems();
	}

	/**
	 * Assigning values to elements to use them as ref
	 */
	private menu!: HTMLElement;

	/**
	 * The aria-live region that will be updated with the selected menu item.
	 */
	private ariaLiveRegion!: HTMLElement;

	render() {
		return (
			<nav
				role="navigation"
				class={
					this.menuIsOpen ? 'ontario-application-navigation ontario-navigation--open' : 'ontario-application-navigation'
				}
				id="ontario-application-navigation"
				ref={(el) => (this.menu = el as HTMLElement)}
			>
				<div class="ontario-application-navigation__container">
					<ul>
						{this.menuItemState.map((item: any) => {
							return this.generateMenuItem(item.href, item.title, item.linkIsActive);
						})}
					</ul>
				</div>
				<div
					id="aria-live-region"
					class="ontario-visually-hidden"
					aria-live="polite"
					ref={(el) => (this.ariaLiveRegion = el as HTMLElement)}
				></div>
			</nav>
		);
	}
}
