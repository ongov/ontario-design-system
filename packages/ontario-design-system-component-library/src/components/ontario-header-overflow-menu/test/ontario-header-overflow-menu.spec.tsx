import { newSpecPage } from '@stencil/core/testing';
import { OntarioHeaderApplicationMenu } from '../ontario-header-overflow-menu';

describe('ontario-header-overflow-menu', () => {
	it('should render without errors', async () => {
		const page = await newSpecPage({
			components: [OntarioHeaderApplicationMenu],
			html: `<ontario-header-overflow-menu menu-items='[{"title":"Link 1","href":"/link-1"},{"title":"Link 2","href":"/link-2"}]'></ontario-header-overflow-menu>`,
		});

		expect(page.root).toBeTruthy();
	});

	it('should render menu items correctly', async () => {
		const menuItems = JSON.stringify([
			{ title: 'Link 1', href: '/link-1', linkIsActive: false },
			{ title: 'Link 2', href: '/link-2', linkIsActive: false },
		]);

		const page = await newSpecPage({
			components: [OntarioHeaderApplicationMenu],
			html: `<ontario-header-overflow-menu menu-items='${menuItems}'></ontario-header-overflow-menu>`,
		});

		expect(page.root).toBeTruthy();
		const menuLinks = page.root?.shadowRoot?.querySelectorAll('a');

		expect(menuLinks?.length).toBe(2);
		expect(menuLinks?.[0]?.textContent).toBe('Link 1');
		expect(menuLinks?.[0]?.getAttribute('href')).toBe('/link-1');
		expect(menuLinks?.[1]?.textContent).toBe('Link 2');
		expect(menuLinks?.[1]?.getAttribute('href')).toBe('/link-2');
	});

	it('should toggle menu visibility when the menuButtonToggled event is received', async () => {
		const menuItems = JSON.stringify([{ title: 'Link 1', href: '/link-1', linkIsActive: false }]);

		const page = await newSpecPage({
			components: [OntarioHeaderApplicationMenu],
			html: `<ontario-header-overflow-menu menu-items='${menuItems}'></ontario-header-overflow-menu>`,
		});

		expect(page.rootInstance).toBeTruthy();
		const component = page.rootInstance!;
		expect(component.menuIsOpen).toBe(false);

		// Simulate the menu button being toggled open
		const event = new CustomEvent('menuButtonToggled', { detail: true });
		window.dispatchEvent(event);
		await page.waitForChanges();

		expect(component.menuIsOpen).toBe(true);
	});

	it('should close menu when clicking outside', async () => {
		const menuItems = JSON.stringify([{ title: 'Link 1', href: '/link-1', linkIsActive: false }]);

		const page = await newSpecPage({
			components: [OntarioHeaderApplicationMenu],
			html: `<ontario-header-overflow-menu menu-items='${menuItems}'></ontario-header-overflow-menu>`,
		});

		expect(page.rootInstance).toBeTruthy();
		const component = page.rootInstance!;
		component.menuIsOpen = true;
		await page.waitForChanges();

		// Simulate clicking outside the menu
		const event = new MouseEvent('click', { bubbles: true });
		document.body.dispatchEvent(event);
		await page.waitForChanges();

		expect(component.menuIsOpen).toBe(false);
	});
});
