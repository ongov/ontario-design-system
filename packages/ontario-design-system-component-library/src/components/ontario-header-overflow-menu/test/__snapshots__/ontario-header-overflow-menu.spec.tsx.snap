// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ontario-header-overflow-menu should render menu items correctly 1`] = `
<ontario-header-overflow-menu menu-items="[{&quot;title&quot;:&quot;Link 1&quot;,&quot;href&quot;:&quot;/link-1&quot;,&quot;linkIsActive&quot;:false},{&quot;title&quot;:&quot;Link 2&quot;,&quot;href&quot;:&quot;/link-2&quot;,&quot;linkIsActive&quot;:false}]">
  <template shadowrootmode="open">
    <nav class="ontario-application-navigation" id="ontario-application-navigation" role="navigation">
      <div class="ontario-application-navigation__container">
        <ul>
          <li>
            <a class="ontario-menu-item" href="/link-1">
              Link 1
            </a>
          </li>
          <li>
            <a class="ontario-menu-item" href="/link-2">
              Link 2
            </a>
          </li>
        </ul>
      </div>
      <div aria-live="polite" class="ontario-visually-hidden" id="aria-live-region"></div>
    </nav>
  </template>
</ontario-header-overflow-menu>
`;
